libstdcpp/unicode-utf-iterator.cpp

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) 2024, Paul Dreik
+// Licensed under Boost software license 1.0
+// SPDX-License-Identifier: BSL-1.0
+
+#include <cstddef>
+#include <cstdint>
+#include <span>
+#include <string_view>
+
+#include <bits/unicode.h>
+
+#include "fuzzcombine.h"
+
+namespace {
+template<typename CharType, typename View>
+constexpr inline auto
+make_view(View&& view)
+{
+  return std::__unicode::_Utf_view<CharType, View>{ std::forward<View>(view) };
+}
+
+/*
+ * this iterates over data, interpreting it as utf data.
+ * the point is to exercise std::__unicode::_Utf_view for a variety
+ * of input and output types
+ */
+template<typename CharType>
+[[clang::optnone]] int
+blah(auto data)
+{
+  auto v = make_view<CharType>(data);
+
+  [[maybe_unused]] auto count = 0u;
+  [[maybe_unused]] unsigned int tmp{};
+  for (auto e : v) {
+    tmp += e;
+    ++count;
+  }
+
+  // could anything be asserted by count vs. data.size()?
+  //  assert(count <= data.size());
```

Comment:
This will work when the output type is `char32_t`, but not otherwise.

For 32-bit input and `char32_t` output you should get `count == data.size()`, because either every input character is a valid UTF-32 code point that can be returned as `char32_t`, or it's invalid and will be returned as the `U'\uFFFD'` replacement character. Either way, it's 1 input to 1 output.

For 16-bit and 8-bit input then `count < data.size()` should be true. Some 16- or 8-bit input values represent a single code point, and so get returned as a `char32_t` output value. But some 16-bit input values are part of a surrogate pair and so two 16-bit code units are consumed to produce a single UTF-32 output. And some 8-bit input values are part of a multibyte sequence, in which case up to 4 bytes can be consumed to produce a single UTF-32 output. In all cases, `count < data.size()` holds.

But when the _output_ is 8-bit or 16-bit, then you might get multiple output values for a single input. e.g. the input range `U"£"` consists of a single `char32_t` code unit, but is encoded as two bytes in UTF-8. Other code points require three or four bytes. So `count >= data.size()`. Similarly, any code point that doesn't fit in a single UTF-16 code unit will get encoded as a surrogate pair, so one 32-bit input value can produce up to two 16-bit output values. And 16-bit inputs with 8-bit outputs can also give an output count higher than the input size.

I think it's accurate to say that if sizeof(input) == sizeof(output) then `count <= data.size()` is true ... but I'd have to think about it further. Even in that case it's not the case that `count == data.size()` because invalid UTF-8 sequences can be transformed to U+FFFD which requires three bytes, so if the input consists of four bytes, the output will be smaller.

Given the complexity of figuring out what the valid inequalities are for different sized inputs and outputs, it might be best to not assert anything here, at least for now.

Reply:
I put this comment into the fuzzer along with checks. The check in the second last paragraph does not pass and I do not know enough about unicode to provide any value to this conversation.... I left it commented out.